File: Interogate.cpp
#include "Interogate.h"

#include <vector>
#include <string>
#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

//mkfifo
#include <sys/types.h>
#include <sys/stat.h>
using namespace std;

#define PATHSIZE 255

FILE* Interogate::resultPipe;
FILE* Interogate::requestPipe;

float Interogate::getJointProbability(std::vector<std::string> &phrase)
{
    std::string strPhrase = "";
    for(unsigned int i = 0; i < phrase.size() - 1; i++)
    {
        strPhrase += phrase[i] + " ";
    }
    strPhrase += phrase[phrase.size() - 1];

    return Interogate::getJointProbability(strPhrase);
}

float Interogate::getJointProbability(std::string phrase)
{
    //trimit request
    phrase += "\n";
    fputs(phrase.c_str(), requestPipe);
    fflush(requestPipe);

    //primesc rezultatul
    char buffer[20];
    fgets(buffer, 20, resultPipe);
    //puts(buffer);

    float result = atof(buffer);
    return result;
}

std::vector<float> Interogate::getJointProbabilities(std::vector<std::string> &phrases)
{
    std::vector<float> result;
    char buffer[100];

    /* Protocol:
     * send a unique string: "Interogate please start buffering\n"
     * send all ngrams separated by "\n"
     * send a unique string: "Interogate please send queries"
     */
    sprintf(buffer, "Interogate please start buffering\n");
    fputs(buffer, requestPipe);
    printf("am trimis interogate pls start\n");
    for(unsigned int i = 0; i < phrases.size(); i++)
    {
        strcpy(buffer, (phrases[i] + "\n").c_str());
        printf("sending %s\n", buffer);
        fputs(buffer, requestPipe);
    }
    fputs("Interogate please send queries\n", requestPipe);
    printf("am trimis interogate pls send\n");

    result.clear();
    for(unsigned int i = 0; i < phrases.size(); i++)
    {
        printf("incep fgets\n");
        fgets(buffer, 100, resultPipe);
        printf("termin fgets\n");
        printf("deb %s\n", buffer);
        result.push_back(atof(buffer));
    }

    return result;
}

void Interogate::Init()
{
    //printf("init 1\n");
    mkfifo("/tmp/ngramfifo", 0666);
    mkfifo("/tmp/ngramfiforeq", 0666);
    //pid_t pidParent = getpid();
    //printf("1 pid %d\n", pidParent);
    pid_t pid = fork();
    //pid_t pid = getpid();
    //printf("2 pid %d\n", pid);
    //if(pid != pidParent)

    //unnamed pipes
    //each pipe is: input | output
    int preq[2], pres[2];
    pipe(preq);
    pipe(pres);

    if(pid == -1)
    {
        printf("ERROR starting the Interogate.py process\n");
        exit(0);
    }
    if(pid == 0)
    {
        //child process
        /*
        char *path = (char*) malloc(PATHSIZE*sizeof(char));
        getcwd(path, PATHSIZE);
        strcat(path, "/Interogate.py");
        int ret = execlp(path, "Interogate.py", NULL);
        */
        close(preq[0]);
        close(pres[1]);
        char spreq[10], spres[10];
        sprintf(spreq, "%d", preq[1]);
        sprintf(spres, "%d", pres[0]);
        int ret = execlp("python", "python", "Interogate.py", spreq, spres, NULL);
        printf("execlp a returnat %d\n", ret);
    }
    //printf("init 2\n");
    requestPipe = fopen("/tmp/ngramfiforeq", "w");
    //requestPipe = fdopen(preq[0], "w");
    if(requestPipe == NULL)
    {
        printf("ERROR opening request pipe\n");
    }
    //printf("init 3\n");
    resultPipe = fopen("/tmp/ngramfifo", "r");
    //resultPipe = fdopen(pres[1], "r");
    if(resultPipe == NULL)
    {
        printf("ERROR opening result pipe\n");
    }
    //printf("init 4\n");
}

void Interogate::Finalize()
{
    fputs("Interogate please exit\n", requestPipe);
    fclose(resultPipe);
    fclose(requestPipe);
}

File: Interogate.py
#! /usr/bin/python

import MicrosoftNgram
import sys
import io
import time

E_VALUE = 2.71828

lookup = MicrosoftNgram.LookupService()

fdreq = int(sys.argv[1])
fdres = int(sys.argv[2])
print 'am primit fd-urile pipurilor anonime: ' + str(fdreq) + ' ' + str(fdres)

#input = fdopen(fdreq, "r");
#out = fdopen(fdres, "w");

out_deb = open("Interogate.out", "w", 0);
out_deb.write('am deschis iesirea\n')
print 'am deschis iesirea\n'

input = open("/tmp/ngramfiforeq", "r")
out_deb.write('am deschis input\n')

out = open("/tmp/ngramfifo", "w", 0);
out_deb.write('am deschis output\n')

buffer = []
buffering = False

while (1):
    line = input.readline(50)
    out_deb.write('am citit ceva de la input\n')
    if line:
        out_deb.write('am primit ' + line)
        if (line == 'Interogate please exit\n'):
            input.close()
            out.close()
            exit(0)
        elif (line == 'Interogate please start buffering\n'):
            out_deb.write('Incep sa fac buffer\n')
            buffer = []
            buffering = True
            continue
        elif (line == 'Interogate please send queries\n'):
            out_deb.write('Am terminat de facut buffer\n')
            result = lookup.GetJointProbabilities(buffer)
            result_final = ""
            for nr in result.split('\n'):
                result_final = result_final + str(E_VALUE**(float(nr))) + "\n"
            out_deb.write("result " + result_final)
            out.write(result_final)
            out.flush()
            out_deb.write('am trimis rezultat final\n')
            buffer = []
            buffering = False
            continue

        if buffering:
            '''remove the '\n' from the end of the line'''
            buffer.append(line[:-1])
            continue
        result = str(E_VALUE**lookup.GetJointProbability(line))
        out_deb.write('am calculat rezultatul ' + result + ' pentru ' + line + '\n')
        out.write(result + '\n')
        out.flush()
    else:
        time.sleep(1)


out.close()
